File: golang/1249_minimum_remove_to_make_valid_parentheses.go
package main

import "fmt"

func main() {
	fmt.Println(minRemoveToMakeValid("lee(t(c)o)de)"))
}

func minRemoveToMakeValid(s string) string {
	stack := []byte{}
	open := 0
	for i := 0; i < len(s); i++ {
		if s[i] == ')' {
			if open > 0 {
				open--
				stack = append(stack, s[i])
			}
		} else if s[i] == '(' {
			open++
			stack = append(stack, s[i])
		} else {
			stack = append(stack, s[i])
		}
		fmt.Printf("open: %d \n", open)
		fmt.Printf("stack: %v \n", stack)

	}
	fmt.Println("================")

	res := make([]byte, len(stack)-open)
	p := len(res) - 1
	fmt.Println(res)
	fmt.Println(p)
	for i := len(stack) - 1; i >= 0; i-- {
		if stack[i] == '(' {
			if open > 0 {
				open--
				continue
			}
			res[p] = '('
			p--
		} else {
			res[p] = stack[i]
			p--
		}
		fmt.Printf("i: %d \n", i)
		fmt.Printf("p: %d \n", p)
		fmt.Printf("open: %d \n", open)
		fmt.Printf("res: %v \n", res)
	}
	return string(res)
}

File: python/1249-Minimum-Remove-to-Make-Valid-Parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        res = []
        first = []
        l = 0
        total = 0
        for c in s:
            if c == '(':
                l += 1
                total += 1
            if c == ')':
                if l == 0:
                    continue
                l -= 1
                print(l)
            first.append(c)
        print(first)
        removed = total - l
        print(removed)
        for c in first:
            if c == '(':
                removed -= 1
                if removed < 0:
                    continue
            res.append(c)
        return "".join(res)


if __name__ == '__main__':
    # begin
    s = Solution()
    print(s.minRemoveToMakeValid("lee(t(c)o)de)"))
